File: homeworks/homework1/22002105_Gokmen_AhmetBerke.rb
# Declare/create an empty list
empty_list_1 = Array.new(0)
empty_list_2 = []
puts "empty_list_1:", empty_list_1
puts "empty_list_2:", empty_list_2

# Initialize a list with some values
list = ['Apple', 'Avocado', 'Dragonfruit', 'Grapefruit', 'Grapes', 'Hazelnut', 'Kiwi', 'Lime', 'Tangelo']
another_list = Array.new(6, "Apple")
puts "list:", list
puts "another_list:", another_list

# Check if the list is empty or not
is_empty_1 = list.empty?
is_empty_2 = empty_list_1.eql? []
puts "is_empty_1:", is_empty_1, "is_empty_2:", is_empty_2

# Add a new element to a list
list << "Raspberries" # add Raspberries to the array
list.push("Cherries") # add Cherries to the array
list.unshift("Watermelon") # add Watermelon to the beginning
puts "list after adding items:", list

# Check if a particular element exists in the list
does_Raspberries_exist = list.include?("Raspberries") # check if 5 exists in the list
puts "does_Raspberries_exist:", does_Raspberries_exist

# Remove a particular element from the list
list.delete_at(0) # delete item at index 0
list.delete("Dragonfruit") # delete 10 if it exists
puts "list after deleting items:", list

# Get the head and the tail of a list
head1, *tail1 = list # same as python :)
head2, tail2 = list[0], list[1..-1] # another way
puts "head1:", head1, "tail1:", tail1

# Print all of the elements in the list
puts "list:", list # simply prints all of the elements line by line
list.each {|e| print e, " "}; print "\n" # another way of printing them in one line

File: homeworks/homework1/22002105_Gokmen_AhmetBerke.lua
function PrintArr(arr) for _, e in pairs(arr) do io.write(e) io.write(" ") end io.write("\n") end

-- Declare/create an empty list
Empty_arr = {}
print("Empty list: ")
PrintArr(Empty_arr)

-- Initialize a list with some values
Arr = {'Apple', 'Avocado', 'Dragonfruit', 'Grapefruit', 'Grapes', 'Hazelnut', 'Kiwi', 'Lime', 'Tangelo'}

print("List with items: ")
PrintArr(Arr)

-- Check if the list is empty or not
Is_empty = #Arr == 0
print("is_empty:", Is_empty)

-- Add a new element to a list
Arr[#Arr + 1] = "Pomegranate" -- first way
table.insert(Arr, "Raspberries") -- another way
print("After adding elements: ")
PrintArr(Arr)

-- Check if a particular element exists in the list
function Exists(arr, el) -- could not figure out another or one liner way of doing this
  for _, e in ipairs(arr) do
    if e == el then
      return true
    end
  end
  return false
end

Does_Pomegranate_exist = Exists(Arr, "Pomegranate")
Does_Melon_exist = Exists(Arr, "Melon")
print("Does_Pomegranate_exist:", Does_Pomegranate_exist, "Does_Melon_exist:", Does_Melon_exist)

-- Remove a particular element from the list
table.remove(Arr, 5)
print("After removing element at index 5 (indexes start at 1): ")
PrintArr(Arr)

Temp = {}
for _, e in pairs(Arr) do if e ~= "Kiwi" then Temp[#Temp + 1] = e end end
Arr = Temp
print("After removing element Kiwi (filtering): ")
PrintArr(Arr)

-- Get the head and the tail of a list
Head, Tail = Arr[1], {} -- in this language, it's really hard to do an inplace filtering or slicing. Thus I had to create a new array
for _, e in pairs(Arr) do if e ~= Arr[1] then Tail[#Tail + 1] = e end end
print("Head:", Head)
io.write("Tail: ") PrintArr(Tail)

-- Print all of the elements in the list
PrintArr(Arr)
